File: ios-code-challenge/Controllers/DetailViewController.m
//
//  DetailViewController.m
//  ios-code-challenge
//
//  Created by Dustin Lange on 1/20/18.
//  Copyright © 2018 Dustin Lange. All rights reserved.
//

#import "DetailViewController.h"

@interface DetailViewController ()

@property (nonatomic, strong) UIBarButtonItem *favoriteBarButtonItem;
@property (nonatomic, assign, getter=isFavorite) BOOL favorite;

@end

@implementation DetailViewController

- (void)configureView
{
    // Update the user interface for the detail item.
    if (self.detailItem) {
        self.detailDescriptionLabel.text = [self.detailItem description];
    }
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    [self configureView];
    
    self.navigationItem.rightBarButtonItems = @[self.favoriteBarButtonItem];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)updateFavoriteBarButtonState
{
    self.favoriteBarButtonItem.image = (self.isFavorite ? [UIImage imageNamed:@"Star-Filled"] : [UIImage imageNamed:@"Star-Outline"]);
}

#pragma mark - Managing the detail item
- (void)setDetailItem:(NSDate *)newDetailItem {
    if (_detailItem != newDetailItem) {
        _detailItem = newDetailItem;
        
        // Update the view.
        [self configureView];
    }
}

#pragma mark - Button Events
- (void)onFavoritBarButtonSelected:(id)sender
{
    _favorite = !self.isFavorite;
    [self updateFavoriteBarButtonState];
}

#pragma mark - Properties
- (UIBarButtonItem *)favoriteBarButtonItem
{
    if(_favoriteBarButtonItem == nil) {
        _favoriteBarButtonItem = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"Star-Outline"]
                                                                  style:UIBarButtonItemStylePlain
                                                                 target:self
                                                                 action:@selector(onFavoritBarButtonSelected:)];
    }
    
    return _favoriteBarButtonItem;
}

@end

File: ios-code-challenge/Controllers/DetailViewController.swift
//
//  DetailViewController.swift
//  ios-code-challenge
//
//  Created by Joe Rocca on 5/31/19.
//  Copyright © 2019 Dustin Lange. All rights reserved.
//

import UIKit

class DetailViewController: UIViewController {

    @IBOutlet weak var thumbnailImageView: UIImageView!
    @IBOutlet weak var nameLabel: UILabel!
    @IBOutlet weak var categoriesLabel: UILabel!
    @IBOutlet weak var ratingLabel: UILabel!
    @IBOutlet weak var reviewCountLabel: UILabel!
    @IBOutlet weak var priceLabel: UILabel!
    
    lazy private var favoriteBarButtonItem: UIBarButtonItem = UIBarButtonItem(image: UIImage(named: "Star-Outline"), style: .plain, target: self, action: #selector(onFavoriteBarButtonSelected(_:)))

    @objc var detailItem: YLPBusiness?
    
    private var _favorite: Bool = false
    private var isFavorite: Bool {
        get {
            return _favorite
        } 
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        configureView()
        navigationItem.rightBarButtonItems = [favoriteBarButtonItem]
    }
    
    private func configureView() {
        guard let business = detailItem else { return }
        nameLabel.text = business.name
        categoriesLabel.text = "Categories: \(business.categories)"
        ratingLabel.text = "Rating: \(String(business.rating))"
        reviewCountLabel.text = "Number of reviews: \(String(business.reviewCount))"
        priceLabel.text = "Price: \(business.price)"
        setImage(with: business.imageUrlString)
    }
    
    func setBusiness(with newBusiness: YLPBusiness) {
        guard detailItem != newBusiness else { return }
        detailItem = newBusiness
        configureView()
    }

    private func setImage(with imageUrlString: String) {
        thumbnailImageView.contentMode = .scaleAspectFit
        thumbnailImageView.clipsToBounds = true
        guard let imageUrl = URL(string: imageUrlString),
            let data = try? Data(contentsOf: imageUrl) else { return }
        DispatchQueue.main.async {
            self.thumbnailImageView.image = UIImage(data: data)
        }
    }
    
    private func updateFavoriteBarButtonState() {
        favoriteBarButtonItem.image = isFavorite ? UIImage(named: "Star-Filled") : UIImage(named: "Star-Outline")
    }
    
    @objc private func onFavoriteBarButtonSelected(_ sender: Any) {
        _favorite.toggle()
        updateFavoriteBarButtonState()
    }
}
